lisp/core.slosh

```diff
@@ -655,6 +673,87 @@ Section: core
 #%
 Asserts the value is an error
 
-Section: core
+Section: test
 %#
 (defmacro test::assert-error (& body) `(assert-error ~@body))
+
+
+#%
+Usage: (apply function arg* list)
+
+Call the provided function with the supplied arguments, last is a list that will be expanded.
+
+Section: core
+
+Example:
+(def test-apply-one (apply str "O" "NE"))
+(test::assert-equal "ONE" test-apply-one)
+(test::assert-equal 10 (apply + 1 2 7))
+%#
+(defmacro apply (a-fn & rst)
+    `(~a-fn ~@rst))
+
+#%
+Usage: (identity arg)
+
+Identity function.
+
+Section: core
+
+Example:
+
+(assert-equal 0 (identity 0))
+%#
+(defn identity (x) x)
+
+#%
+Replaces all instances of old-item in lst with new-item. If last argument
+passed in is keyword :first only the first instance of old-item will be
+replaced.
+
+Section: core
+
+Example:
+(let (lst (list 1 2 3 4 5))
+    (test::assert-equal (list 1 2 2 4 5) (nsubstitute! lst 3 2))
+    (test::assert-equal (list 1 2 2 4 5) lst)
+    (test::assert-equal (list 1 3 2 4 5) (nsubstitute! lst 2 3 :first)))
+%#
+(defn nsubstitute!
+    (lst old-item new-item & mods)
+    (let (early-return (not (nil? mods)))
```

Comment:
I wasn't able to make this check

`(early-return (in? mods :first)`

when I print out mods when called directly from nsubstitute! like:
```
(nsubstitute! lst 2 3 :first)
```
when the implementation starts like this:
```
(defn nsubstitute!
    (lst old-item new-item & mods)
    (prn "mods: " mods ", type: " (type mods))
    (let (early-return (in? mods :first))
    ...
```
it prints:
`mods: (#<Undefined>), type: :Pair`

when called directly running the shell it prints:
`mods: (:first), type: :Pair`

and I do not know why.